### ISSUE TITLE
Warn if in lockdown mode

### ISSUE THREAD

**Issue by mmarchini** [iovisor/bpftrace#925]
> his restricts a number of kernel APIs when you boot from UEFI secure boot with a signed kernel. Those APIs include kprobe and bpf.

If a version of lockdown is going to end up in mainline it would be nice if we could detect this mode and generate an appropriate error, like we do for running without root.

_Originally posted by @fbs in https://github.com/iovisor/bpftrace/issues/853#issuecomment-521531382_

**Comment by fbs** [comment]
linking to https://github.com/iovisor/bcc/issues/2565

**Comment by dailyherold** [comment]
My freshly installed 19.10 laptop with `5.3.0-29-generic` kernel had issues with this. Thanks to @psanford in #853 disabling lockdown mode fixed my issues both with bpftrace and bcc (iovisor/bcc#2676). Definitely think we may need to escalate this if all 19.10 users are going to be experiencing the same thing with a fresh install.

**Comment by fbs** [comment]
@dailyherold which bpftrace version were you using? Do you  have a  stacktrace of when it happens? Bpftrace shouldn't segv when it failes to attach code / create maps.

**Comment by dailyherold** [comment]
@fbs I was getting segfaults for any attempt, even hello world, see https://github.com/iovisor/bpftrace/issues/664#issuecomment-583803813. Those results were built off 2667b8a2e1c22613e731c7b68d66e742a58f2249. I can re-enable lockdown and get some more messages from dmesg, or if there's any other more verbose logging/stacktraces that would be helpful? I also have realized desktop ubuntu isn't the primary use case for these tools, so I don't think "escalating" is really necessary. But I do imagine that people picking up the BPF book (like me!) might be running newer distros and wanting to mess around on their local machine while not at work.

**Comment by fbs** [comment]
It seems like  lockdown mode is  only enabled on uefi systems with secure boot which makes it a bit tricky for me to tests with. If you can make a gdb stacktrace that would be quite helpful as the segv doesn't tell me much

```
[12102.514807] bpftrace[16858]: segfault at 5 ip 00007fe60e4f4ba5 sp 00007fff9d4dc808 error 4 in libc-2.30.so[7fe60e38e000+178000]
[12102.514822] Code: 00 00 0f 1f 00 31 c0 c5 f8 77 c3 66 2e 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 89 f9 48 89 fa c5 f9 ef c0 83 e1 3f 83 f9 20 77 2b <c5> fd 74 0f c5 fd d7 c1 85 c0 0f 85 eb 00 00 00 48 83 c7 20 83 e1
```

**Comment by brendangregg** [comment]
I've written a note about lockdown here: https://github.com/iovisor/bcc/issues/2565#issuecomment-584476552

We need a warning ASAP that links to https://github.com/iovisor/bpftrace/blob/master/INSTALL.md#disable-lockdown , and add that section to INSTALL.md with a current summary of what's going on and how to stop it. It should also ask people to complain to their OS vendors (or switch OS vendors.)

**Comment by fbs** [comment]
I found a  ubuntu 19.10 system with secure boot which I can use for testing. Looking through the patch I  don't see a way to detect whether lockdown is enabled by scanning through /proc /sys sysctl etc. The best options I've found are:

- kprobes fail (attach a dummy program that exits)
- BPF_MAP_TYPE_PERF_EVENT_ARRAY is blocked

Both failing with EPERM should be a pretty good indication that lockdown is enabled. Or atleast  that bpftrace won't work in the current state of the system.
Maybe we need to check for cap_sys_admin too, to rule out containers/namespaces? I'm not sure how to best detect that.

**Comment by fbs** [comment]
So ubuntu 19.10 still has old patches which fully disable the bpf syscall. Which means that bpf on a locked down 19.10 is completely useless. Even the creation of an array map fails. It also doesn't log anything in dmesg.

I couldn't reproduce the segv @dailyherold  was seeing, pretty certain his issue have been fixed in master (e.g. #774 ).

I've reinstalled my system to fedora 31, which appears to have the upstream lockdown stuff. Haven't had  time to dig into it yet but on first glance I noticed tracefs was missing and that it actually logs stuff  to  dmesg.
